### DISCUSSION TITLE
Spline support

### DISCUSSION THREAD

**Discussion by Aggror** [stride3d/stride#1212] [Contributors]
I have been working on Splines (a bit on and off) and I would like to have some feedback on my approach so far:

For the base of Spline support I want to have the following options:

Components:
- Splinecomponent 
  - has various debug options
  - Holds a reference to spline node components
  - Loop option
- SplineNodeComponent
  - Calculates the spline
  - Segments (Minimal of 2, in case you want a straight line)

Requirements:
- Spline should be visible in the editor
- Moving the splinenodes, segments count, or changing the spline control tangents should update the spline.

Current shortcomings:
- Loading a scene with a spline setup, doesn't properly display the spline. Updating the spline manually (moving/updating any component fixes the spline gizmo)
- Nice to have: The node tangents are stored in properties on the spline node component. Ideally I would like to have 2 additional translation gizmos visible in the editor that allow for easier manipulation of the tangent.
  -SplineNodeGizmo and SplineNode, would be the place were this is introduced.  
- Current gizmo is a flat line (like a thin road.) Which means that at some angles you can't always see the spline debug mesh properly.

Here is a video demonstrating above: https://www.youtube.com/watch?v=CYJ5Jf370ns

![image](https://user-images.githubusercontent.com/3499539/145029380-26556285-1cdf-49f2-b629-1eb80a1486dc.png)

Stride repo: https://github.com/Aggror/Stride/tree/SplineBase
Code is not final, nor is it properly formatted with the coding guidelines of Stride yet. I am trying to work out the functionality first.

Base test project: https://github.com/Aggror/SplineTools

Future components:
- Spline traveller: moving, rotating along the spline (ongoing, early wip)
- Spline decorator: (ongoing, early wip, https://www.youtube.com/watch?v=igm3s1NzNSo)
- Spline mesh generator: various meshes that can be generated or even take a base model, to generate mesh along the spline (road, cable, traintracks, rollercoaster etv)

**Comment by MechWarrior99** [comment]
Its looking nice! However, perhaps it would be better to have the spline as a basic standalone class that components can then use. This would let it be more flexible for future tools/components. And a standard component that uses it could be provided as well like you have now.

An example use-case might be a road system. You most likely would want to have multiple splines but not want to have to add a new entity/component for each one.

**Reply by Aggror** [reply]
Thanks for the reply. A single curve is already in the BezierCurve class. The Spline Component however will utilize the existing spline class, which was initially abandoned to make things easier during prototyping. The endgoal is a single (code only) sync script where you can create  `new Spline() `with `new SplineNode()`

Roads are part of the spline mesh generators that have yet to come.

**Comment by manio143** [comment]
I'm thinking about the approach you took with spline nodes, and at the moment I think having child entities is the simplest option. In that case rather than keeping a list with references you can just apply the ordering of the children.

A more advanced system could keep everything within the spline component, such that you would have a list of nodes (TRS + tangents) and maintain node selection within the spline gizmo - this would require the gizmo to include both TRS controller for tangent balls and for the nodes.

On the editor weird behavior messing up your spline - no clue - my constraint gizmo doesn't show up when you add a component straight away, but selecting another entity and reselecting makes it appear.

**Reply by Aggror** [reply]
> I'm thinking about the approach you took with spline nodes, and at the moment I think having child entities is the simplest option. In that case rather than keeping a list with references you can just apply the ordering of the children.

Thanks for the reply @manio143. I have used the 'Child' approach in the early beginning but stopped using it because I had thought that I would have more control over lists and reordering lists in the editor. Turns out that those controls are actually quite limited. But another downside of using children, is that the parent entity doesn't know about those children if they were added to the scene and given a spline node component. The parent would have to check every update if the children with the spline node component would have changed since the previous update. I didn't quite like that at the time, so I went with the List option. I think the most viable alternative (if I were to implement it), would be that every node component refers to another node component. The Spline is no longer the controller of the spline. Both methods have their ups and downs. 

> A more advanced system could keep everything within the spline component, such that you would have a list of nodes (TRS + tangents) and maintain node selection within the spline gizmo - this would require the gizmo to include both TRS controller for tangent balls and for the nodes.

This sounds like it is a little bit more than I can chew. I experimented with instantiating the translation gizmo's in various locations, to see I could get something that I could work with, but all of it ended with crashes and confusing errors. 

> On the editor weird behavior messing up your spline - no clue - my constraint gizmo doesn't show up when you add a component straight away, but selecting another entity and reselecting makes it appear.

Yeah, there is something that doesn't update the scene directly but I can't put my finger on it. I also had issues with components being updated several times in a row, which triggered al kinds of logic. This is sort of the nature of filling in the data contracts I guess.

**Comment by Aggror** [comment]
Quick update:

-  I did some more testing with the following scenarios: Spline by code (instead of editor) and Closest point demo
- There still is this weird bug were I sometimes seem to be getting a 'ghost' spline. It has cost me a lot of time finding out were it is coming from, but no result yet.
- Not all spline options update the spline renderer properly. I have been toying around with properties, nested properties/datacontract properties, but it is just not 100% foolproof. At this point I have tried so many things, that I also kind of lost track of what the proper way is to achieve my goal. 
- Given up on the 'Tangent gizmo'. It is just too complex for me too handle.

**Spline by code**
A simple script that lets you create a spline by code and visualise the spline and/or boundingboxes during runtime.

![image](https://user-images.githubusercontent.com/3499539/149405247-b564bd0e-de8c-4ea5-8a49-4b37413af7e5.png)

`public override void Start()
{
	var nodePositions = new Vector3[3]
	{
		new Vector3(4, 1, 0),
		new Vector3(0, 2, 2),
		new Vector3(-2, 1, -1)
	};

	var tangents = new Vector3[6]
	{
		new Vector3(0, 1, 4),   //Node 1 - out
		new Vector3(-1, 0, -3), //Node 1 - in
		new Vector3(4, 1, -2), //Node 2 - out
		new Vector3(-3, 0, 2),  //Node 2 - in
		new Vector3(-5, -1, -1), //Node 3 - out
		new Vector3(4, 0, 0)    //Node 3 - in
	};

	splineComponent = new SplineComponent();
	Entity.Add(splineComponent);

	for (int i = 0; i < nodePositions.Length; i++)
	{
		var nodeEntity = new Entity(nodePositions[i]);
		var nodeComponent = new SplineNodeComponent(50, tangents[i*2], tangents[i*2+1]);
		nodeEntity.Add(nodeComponent);

		Entity.AddChild(nodeEntity);
		splineComponent.Nodes.Add(nodeComponent);
	}

	// We use a spline renderer if we want to show our spline in the game
	splineComponent.SplineRenderer.SegmentsMaterial = splineMaterial;
	splineComponent.SplineRenderer.Segments = true;
	splineComponent.SplineRenderer.BoundingBoxMaterial = boundingBoxMaterial;
	splineComponent.SplineRenderer.BoundingBox = true;
}`

**Closest point on spline demo**
A simple script that allows the movement of a spere. The same script also checks the referenced spline for its nearest point and places a red sphere on that position. You can simply retrieve closest point information from the spline by providing a world position and the spline class gives you some info in return like position, distance, closest nodes etc.

`var closestPositionInfo = spline.GetClosestPointOnSpline(Entity.Transform.WorldMatrix.TranslationVector);`

https://user-images.githubusercontent.com/3499539/149404742-a14fce37-a324-4c1b-a6dd-e205700bdb7f.mp4

**Reply by VaclavElias** [reply]
It looks nice! When I finish my code-only approach, I will play with this as well in the code-only project :)

**Comment by Aggror** [comment]
First PR and recording out the door: https://github.com/stride3d/stride/pull/1287

**Reply by VaclavElias** [reply]
Nice PR, also I can see what needs to be added to the Engine if you add such a new feature.

**Comment by CursiveCrow** [comment]
Has there been any movement on this? I'm evaluating stride and our current project depends almost entirely on splines.

**Reply by Aggror** [reply]
I have merged the .net 8 changes and applied most of the performance suggestions and other feedback on the PR. Once the holiday period starts I would like to have the base PR ready. I would have to ask another dev besides Manio to have a look as well.

I have lots of additional components that would like to port that I already use in my own project. Things like a spline decorator, mesh generators

**Reply by Aggror** [reply]
Here is some ongoing testing if you are interested: 
https://youtu.be/N8vkfUdy4E0
